qiskit/transpiler/passes/basis/basis_translator.py

```diff
@@ -154,7 +154,7 @@ def run(self, dag):
 
                 if (len(bound_target_dag.op_nodes()) == 1
                         and len(bound_target_dag.op_nodes()[0].qargs) == len(node.qargs)):
-                    dag_op = bound_target_dag.op_nodes()[0].op
+                    dag_op = bound_target_dag.op_nodes()[0].op.copy()
```

Comment:
Can this copy be done inside `DAGCircuit.substitute_node`? This will also fix this behavior for other places where `substitute_node` is called.

It also seems likely this will end up slower (because we're introducing many more copies). Can this be mitigated by only performing the copy when `condition is not None`?

Reply:
It could be, but I think the 2 gates with the same instance may only be a problem in BasisTranslator, so if it's put in substitute_node there might be some needless copies. Don't know if the problem exists elsewhere.

I understand there's a performance issue, but if you skip the copy for some cases, you will be sending 2 or more gates down to later passes that are the same instance. Seems like this could cause a difficult to find problem, but don't have any specific case.

Reply:
Agree that this is a problem that comes up mostly for us in the `BasisTranslator`, but I think we want to guard against this in case any out of tree or future passes use this behavior. This is consistent with e.g. `DAGCircuit.substitute_node_with_dag` ( https://github.com/Qiskit/qiskit-terra/blob/2893fc8/qiskit/dagcircuit/dagcircuit.py#L976 ) and elsewhere that if an object is passed as an argument, it won't be mutated without a copy. These can be minimized if the copy is only made when `condition is not None`.

This is also a good point, and one which we ought to standardize and document more clearly because our position has been slowly shifting. Repeating `Instruction` instances in a circuit should be a supported pattern, and any downstream passes should not be mutating `Instruction` instances in place. (This was already the case for the `BasisTranslator` pass.) The thinking here was that as circuit sizes grow, requiring an individual `Instruction` class instance per instruction will become an avoidable overhead.

Reply:
All makes sense. If a multiple use instance heads downstream, it's up to the downstream pass to do what it needs to if it needs to mutate. I'll go ahead and add the `condition is not None` check before copying.